flink-core/src/main/java/org/apache/flink/api/common/eventtime/WatermarkStrategy.java

```diff
@@ -49,4 +53,158 @@
+	 * access to the metrics system.
+	 *
+	 * <pre>
+	 * {@code WatermarkStrategy<Object> wmStrategy = WatermarkStrategy
+	 *   .forMonotonousTimestamps()
+	 *   .withTimestampAssigner((ctx) -> new MetricsReportingAssigner(ctx));
+	 * }</pre>
+	 */
+	default WatermarkStrategy<T> withTimestampAssigner(TimestampAssignerSupplier<T> timestampAssigner) {
+		checkNotNull(timestampAssigner, "timestampAssigner");
+		return new WithTimestampAssigner<>(this, timestampAssigner);
+	}
+
+	/**
+	 * Creates a new {@link WatermarkStrategy} with the {@link TimestampAssigner} overridden by the
+	 * provided assigner.
+	 *
+	 * <p>You can use this in case you want to specify a {@link TimestampAssigner} via a lambda
+	 * function.
+	 *
+	 * <pre>
+	 * {@code WatermarkStrategy<CustomObject> wmStrategy = WatermarkStrategy
+	 *   .forMonotonousTimestamps()
+	 *   .withTimestampAssigner((event, timestamp) -> event.getTimestamp());
+	 * }</pre>
+	 */
+	default WatermarkStrategy<T> withTimestampAssigner(SerializableTimestampAssigner<T> timestampAssigner) {
+		checkNotNull(timestampAssigner, "timestampAssigner");
+		return new WithTimestampAssigner<>(this, TimestampAssignerSupplier.of(timestampAssigner));
+	}
+
+	/**
+	 * Add an idle timeout to the watermark strategy. If no records flow in a partition of a stream
+	 * for that amount of time, then that partition is considered "idle" and will not hold back the
+	 * progress of watermarks in downstream operators.
+	 *
+	 * <p>Idleness can be important if some partitions have little data and might not have events
+	 * during
+	 * some periods. Without idleness, these streams can stall the overall event time progress of
+	 * the application.
+	 */
+	default WatermarkStrategy<T> withIdleness(Duration idleTimeout) {
+		checkNotNull(idleTimeout, "idleTimeout");
+		checkArgument(!(idleTimeout.isZero() || idleTimeout.isNegative()),
+				"idleTimeout must be greater than zero");
+		return new WithIdlenessStrategy<>(this, idleTimeout);
+	}
+
+	/**
+	 * A {@link WatermarkStrategy} that adds idleness detection on top of the wrapped strategy.
+	 */
+	class WithIdlenessStrategy<T> implements WatermarkStrategy<T> {
+
+		private static final long serialVersionUID = 1L;
+
+		private final WatermarkStrategy<T> baseStrategy;
+		private final Duration idlenessTimeout;
+
+		private WithIdlenessStrategy(WatermarkStrategy<T> baseStrategy, Duration idlenessTimeout) {
+			this.baseStrategy = baseStrategy;
+			this.idlenessTimeout = idlenessTimeout;
+		}
+
+		@Override
+		public TimestampAssigner<T> createTimestampAssigner(TimestampAssignerSupplier.Context context) {
+			return baseStrategy.createTimestampAssigner(context);
+		}
+
+		@Override
+		public WatermarkGenerator<T> createWatermarkGenerator(WatermarkGeneratorSupplier.Context context) {
+			return new WatermarksWithIdleness<>(baseStrategy.createWatermarkGenerator(context),
+					idlenessTimeout);
+		}
```

Comment:
nit: we should add hashCode/equals here, otherwise it is difficult to compare strategies for e.g. for testing.

Reply:
Not sure. Is `equals` and `hashode` well-defined here? These are no classes describing data or state, but classes mainly representing logic.

If testing is the sole purpose, then the right way to go, in my opinion, is writing a `Matcher` or an utility method that tests for the equality. Pushing a not-very-well-defined `equals()` implementation into production code just to simplify tests is not the right way.